### PR TITLE
drivers: can: mcan: Add CAN statistics

### PR DESCRIPTION
**Pull request by gramsay0**
I had a go at adding MCAN statistics, although I ran into a couple of issues:

- I tested a TX with nothing connected to the CAN transceiver and got endless AckError ISRs (>100k in 30 seconds)
  - i.e. shouldn't this stop when the controller enters an error state?
- ~The PSR Protocol Exception Event (PXE) bit never seems to get set~ - I misunderstood what this is used for

I am testing with this MCAN version:
```
can_mcan: IP rel: 3.2.3 08.6.8
```

Any thoughts/ideas on these?

@henrikbrixandersen

### CODE REVIEW

**Comment by gramsay0** [comment]
I tested this on Linux and got the same result, hundreds of thousands of bus-errors:

image

Interestingly, if I disable the transceiver, lower the bitrate or sneeze then the behavior is different (usually the ISR is called just once and the controller goes to bus-off).

image

Does the ISR maybe need a hold-off time to prevent chewing up the entire CPU under certain circumstances?

**Comment by henrikbrixandersen** [comment]
> * I tested a TX with nothing connected to the CAN transceiver and got endless AckError ISRs (>100k in 30 seconds)  
>   * i.e. shouldn't this stop when the controller enters an error state?

No, that is expected. The Bosch M_CAN will attempt retransmission when it sees a missing ACK (and trigger the AckError interrupt). The retransmission also sees a missing ACK and so it continues.

File: drivers/can/can_mcan.c [hunk 1]
```diff
@@ -562,10 +610,19 @@ void can_mcan_line_0_isr(const struct device *dev)
 			LOG_ERR("Access to reserved address");
 		}
 
-		if (ir & CAN_MCAN_IR_MRAF) {
+		if ((ir & CAN_MCAN_IR_MRAF) != 0U) {
 			LOG_ERR("Message RAM access failure");
 		}
 
+		if ((ir & (CAN_MCAN_IR_PEA | CAN_MCAN_IR_PED)) != 0U) {
+			LOG_WRN("Protocol Error");
```
Comment (henrikbrixandersen) [inline drivers/can/can_mcan.c:618]: I don't think a log message is warranted here. That could lead to quite a lot of log messages. This is why the CAN stats was introduced in the first place.

File: include/zephyr/drivers/can/can_mcan.h [hunk 2]
```diff
@@ -127,6 +127,17 @@
...
+enum can_mcan_psr_lec {
+	CAN_MCAN_PSR_LEC_NO_ERROR    = 0,
+	CAN_MCAN_PSR_LEC_STUFF_ERROR = 1,
+	CAN_MCAN_PSR_LEC_FORM_ERROR  = 2,
+	CAN_MCAN_PSR_LEC_ACK_ERROR   = 3,
+	CAN_MCAN_PSR_LEC_BIT1_ERROR  = 4,
+	CAN_MCAN_PSR_LEC_BIT0_ERROR  = 5,
+	CAN_MCAN_PSR_LEC_CRC_ERROR   = 6,
+	CAN_MCAN_PSR_LEC_NO_CHANGE   = 7
+};
```
Comment (henrikbrixandersen) [inline include/zephyr/drivers/can/can_mcan.h:139]: Please avoid converting register bit definitions into enumerations. Nothing prevents multiple bits from being set.

File: drivers/can/can_mcan.c [hunk 3]
```diff
@@ -524,13 +524,61 @@ static void can_mcan_tx_event_handler(const struct device *dev)
...
+
 void can_mcan_line_0_isr(const struct device *dev)
 {
 	const uint32_t events = CAN_MCAN_IR_BO | CAN_MCAN_IR_EP | CAN_MCAN_IR_EW |
 				CAN_MCAN_IR_TEFN | CAN_MCAN_IR_TEFL | CAN_MCAN_IR_ARA |
-				CAN_MCAN_IR_MRAF;
+				CAN_MCAN_IR_MRAF | CAN_MCAN_IR_PEA | CAN_MCAN_IR_PED;
 	struct can_mcan_data *data = dev->data;
-	uint32_t ir;
+	uint32_t ir, reg;
```
Comment (henrikbrixandersen) [inline drivers/can/can_mcan.c:581]: Please honor the coding style of having each variable on a separate line.

File: drivers/can/can_mcan.c [hunk 4]
```diff
@@ -524,13 +524,61 @@ static void can_mcan_tx_event_handler(const struct device *dev)
...
+		break;
+	}
+}
+
+static void can_mcan_psr_update_stats(const struct device *dev, uint32_t psr_reg)
+{
+	/* Reading the lower byte of the PSR register clears the protocol last
+	 * error codes (LEC). To avoid missing errors, this function should be
+	 * called whenever the PSR register is read.
+	 */
```
Comment (henrikbrixandersen) [inline drivers/can/can_mcan.c:561]: Another options could be to introduce a `can_mcan_read_psr()` function which automatically handles this.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Comment by henrikbrixandersen** [comment]
I wonder how you verified this? Did you use an out-of-tree driver front-end? You neglected to update all the in-tree driver front-ends from using `DEVICE_DT_INST_DEFINE()` to using `CAN_DEVICE_DT_INST_DEFINE()`.

**Comment by gramsay0** [comment]
> > * I tested a TX with nothing connected to the CAN transceiver and got endless AckError ISRs (>100k in 30 seconds)
> >   
> >   * i.e. shouldn't this stop when the controller enters an error state?
> 
> No, that is expected. The Bosch M_CAN will attempt retransmission when it sees a missing ACK (and trigger the AckError interrupt). The retransmission also sees a missing ACK and so it continues.

Yes, although I thought the controller would go to bus-off after enough failures.

I tested this with a zero byte CAN payload at 1Mbps, it generated an ISR every 54us until the bus state went to error-passive, and then every 62us forever (without going to bus-off).

Do you think this could overload a device, preventing it from processing anything else (maybe worse with tcan4x5x over SPI)?

File: include/zephyr/drivers/can/can_mcan.h (hunk 2 above)
Comment (gramsay0) [reply include/zephyr/drivers/can/can_mcan.h:139]: Last error code (LEC) is a 3bit integer.
I think an enumeration makes sense here, these are all the possible values

File: drivers/can/can_mcan.c (hunk 4 above)
Comment (gramsay0) [reply drivers/can/can_mcan.c]: Agreed, that is much tidier

File: drivers/can/can_mcan.c [hunk 5]
```diff
@@ -524,11 +524,67 @@ static void can_mcan_tx_event_handler(const struct device *dev)
...
+#ifdef CONFIG_CAN_STATS
+	enum can_mcan_psr_lec lec;
+
+	lec = FIELD_GET(CAN_MCAN_PSR_LEC, *val);
+	can_mcan_lec_update_stats(dev, lec);
+#ifdef CONFIG_CAN_FD_MODE
+	lec = FIELD_GET(CAN_MCAN_PSR_DLEC, *val);
+	can_mcan_lec_update_stats(dev, lec);
+#endif
+#endif
```
Comment (alexanderwachter) [inline drivers/can/can_mcan.c:578]: Please add a comment which #if is closing
```suggestion
#endif /* CONFIG_CAN_STATS */
```

File: drivers/can/can_mcan.c [hunk 6]
```diff
@@ -524,11 +524,67 @@ static void can_mcan_tx_event_handler(const struct device *dev)
...
+		break;
+	case CAN_MCAN_PSR_LEC_NO_ERROR:
+	case CAN_MCAN_PSR_LEC_NO_CHANGE:
+	default:
+		break;
+	}
+}
+#endif
+
+int can_mcan_read_psr(const struct device *dev, uint32_t *val)
```
Comment (alexanderwachter) [inline drivers/can/can_mcan.c:557]:
```suggestion
static int can_mcan_read_psr(const struct device *dev, uint32_t *val)
```

Comment (gramsay0) [reply drivers/can/can_mcan.c]: Thanks @alexanderwachter, fixed

**Comment by henrikbrixandersen** [comment]
> Yes, although I thought the controller would go to bus-off after enough failures.

Do you have `CAN_AUTO_BUS_OFF_RECOVERY=y`?

**Comment by gramsay0** [comment]
> > Yes, although I thought the controller would go to bus-off after enough failures.
> 
> Do you have `CAN_AUTO_BUS_OFF_RECOVERY=y`?

I did try toggling that, same result.

Is auto-recovery implemented for MCAN, it looked like the config only optionally compiled the recover function?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks good! One thing missing is a call to `CAN_STATS_RESET()` in `can_mcan_start()`. See https://github.com/zephyrproject-rtos/zephyr/blob/8e7c7c62b15cbc09633e48f95eec6aa9150a4791/drivers/can/can_mcux_flexcan.c#L314 for inspiration.

**Comment by henrikbrixandersen** [comment]
> Is auto-recovery implemented for MCAN, it looked like the config only optionally compiled the recover function?

That implementation does look a bit fishy. We will need to look into that. From a quick test here, it looks like the controller _is_ going into bus-off, but auto-recovers. I think this explains the jump from 54 usec to 62 usec.

**Comment by gramsay0** [comment]
> That implementation does look a bit fishy. We will need to look into that.

👍

> I think this explains the jump from 54 usec to 62 usec.

I think it's the "8 bit suspend transmission time" when in error-passive state (1Mbps == 1bit/us):
image
Bus-off recovery should be 128*11 = 1408us

**Comment by henrikbrixandersen** [comment]
> I think it's the "8 bit suspend transmission time" when in error-passive state (1Mbps == 1bit/us):

Ahh, right. Good find. I still think it is okay behaviour for now. If you can add the call to reset the statistics on start, I think it's ready to go in.

**Review by henrikbrixandersen — APPROVED** [review]

**Review by martinjaeger — APPROVED** [review]

**Review by alexanderwachter — APPROVED** [review]

**Pull request merged**